design/proposals/namespace_label_authenticator.md

```diff
@@ -0,0 +1,573 @@
+      end
+    end
+  end
+end
+```
+
+### User Interface
+[//]: # "Describe user interface (including command structure, inputs/outputs, etc where relevant)"
+
+Users will interact with this new feature when configuring host identities in
+policy. All other factors in Authenticator setup are unchanged.
+
+```yaml
+- !host
+  id: test-app
+  annotations:
+    authn-k8s/namespace-label: "key1=value1"  
+```
+
+Rancher creates a unique ID for Projects at creation. This ID needs to inform
+the host identity configuration, and can be found either through Rancher's GUI,
+API, or CLI.
+
+```shell-session
+# API call to retrieve the ID for a project
+PROJECT_ID="$(curl -sk -H "Authorization: Bearer ${RANCHER_TOKEN}" \
+  https://rancher.myorg.com/v3/projects?name=${PROJECT_NAME} \
+  | jq '.data[0].id' \
+  | tr -d '"' \
+  | cut -d ":" -f2)"
+```
+
+```shell-session
+# CLI call to retrieve the ID for a project
+rancher login --token ${RANCHER_TOKEN} https://rancher.myorg.com/v3
+PROJECT_ID="$(rancher project ls \
+  | grep ${PROJECT_NAME} \
+  | awk '{print $1}' \
+  | cut -d ":" -f2)"
+```
+
+This value then informs host configuration:
+```yaml
+- !host
+  id: test-app
+  annotations:
+    authn-k8s/namespace-label: "field.cattle.io/projectId=${PROJECT_ID}"
+```
+
+## Design
+[//]: # "Add any diagrams, charts and explanations about the design aspect of the solution. Elaborate also about the expected user experience for the feature"
+
+### Class / Component Diagrams
+[//]: # "Describe classes that are going to be added /changes and their immediate environment. Non-changed classes may be colored differently"
+
+![Updated AuthnK8s Request Validation Process Diagram](./updated-authnk8s-req-validation-process.png)
+
+### Sequence Diagrams
+[//]: # "Describe main flows in system influenced by this design - using sequence diagram UML"
+
+![Updated AuthnK8s Request Validation Sequence Diagram](./updated-authnk8s-req-validation-sequence.png)
+
+## Performance
+[//]: # "Describe potential performance issues that might be raised by the system as well as their mitigations"
+[//]: # "How does this solution affect the performance of the product?"
+
+| **Subject** | **Description** | **Issue Mitigation** |
+|-------------|-----------------|----------------------|
+| **Many Similarly-Labeled Namespaces** | Using Kubernetes' native label selector implementation would mean receiving a list of all similarly-labeled namespaces, and limiting those results further based on a request's origin namespace. For a cluster with many similarly-labeled namespaces, this operation could be costly. | Instead of querying the Kubernetes API for a list of all similarly-labeled namespaces, we can get only the request's origin namespace, and then validate its labels against the configured selector locally. This means Conjur is only ever querying for a single resource instead of a open-ended multitude. |
```

Comment:
Worth mentioning that from a performance standpoint, it's still one more call than all our other types of authentications? (service account, controller, namespace and so on)
Because in others, we query the Pod and we get all the relevant information we need from that object. Here we will need to perform one more call, on the Namespace object of the Pod to get its labels.